### PR TITLE
Add ability to disable the raster thread merger

### PR DESCRIPTION
**Pull request by blasten**
## Description

Reworks dynamic thread merging, so there isn't need to merge threads on `Shell::OnPlatformViewDestroyed()`. https://github.com/flutter/engine/pull/19919/files

## Related Issues

https://github.com/flutter/flutter/issues/57067

## Tests

I added the following tests: Unit tests

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [contributor guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [C++, Objective-C, Java style guides] for the engine.
- [x] I read the [tree hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation.
- [x] All existing and new tests are passing.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[contributor guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[tree hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: shell/common/shell_unittests.cc [hunk 1]
```diff
@@ -597,201 +597,6 @@ TEST_F(ShellTest,
   DestroyShell(std::move(shell));
 }
 
-TEST_F(ShellTest, OnPlatformViewDestroyAfterMergingThreads) {
```
Comment (cyanglaz) [inline shell/common/shell_unittests.cc:600]: Ideally, this test should pass with this change.

File: shell/common/shell_unittests.cc [hunk 2]
```diff
@@ -597,201 +597,6 @@ TEST_F(ShellTest,
...
-      shell->GetTaskRunners().GetRasterTaskRunner()->GetTaskQueueId(),
-      shell->GetTaskRunners().GetPlatformTaskRunner()->GetTaskQueueId()));
-
-  // Validate the platform view can be recreated and destroyed again
-  ValidateShell(shell.get());
-
-  DestroyShell(std::move(shell));
-}
-
-TEST_F(ShellTest, OnPlatformViewDestroyWhenThreadsAreMerging) {
```
Comment (cyanglaz) [inline shell/common/shell_unittests.cc:666]: Same here, this test should still pass

File: shell/common/shell_unittests.cc [hunk 3]
```diff
@@ -597,201 +597,6 @@ TEST_F(ShellTest,
...
-      shell->GetTaskRunners().GetPlatformTaskRunner()->GetTaskQueueId()));
-
-  // Validate the platform view can be recreated and destroyed again
-  ValidateShell(shell.get());
-
-  DestroyShell(std::move(shell));
-}
-
-TEST_F(ShellTest,
-       OnPlatformViewDestroyWithThreadMergerWhileThreadsAreUnmerged) {
```
Comment (cyanglaz) [inline shell/common/shell_unittests.cc:740]: Should also pass.

File: fml/raster_thread_merger.cc [hunk 4]
```diff
@@ -16,16 +16,19 @@ RasterThreadMerger::RasterThreadMerger(fml::TaskQueueId platform_queue_id,
...
 }
 
 void RasterThreadMerger::MergeWithLease(size_t lease_term) {
+  FML_DCHECK(lease_term > 0) << "lease_term should be positive.";
   if (TaskQueuesAreSame()) {
     return;
   }
-
-  FML_DCHECK(lease_term > 0) << "lease_term should be positive.";
+  if (!IsEnabledUnSafe()) {
```
Comment (cyanglaz) [inline fml/raster_thread_merger.cc:29]: Should we move the lock here?

Comment (blasten) [reply fml/raster_thread_merger.cc:29]: Done

File: shell/common/shell_unittests.cc (hunk 1 above)
Comment (blasten) [reply shell/common/shell_unittests.cc:600]: Done.

File: shell/common/shell_unittests.cc (hunk 2 above)
Comment (blasten) [reply shell/common/shell_unittests.cc:666]: Done.

**Comment by blasten** [comment]
I have verified that this patch fixes the reason #20722 was reverted in #20745.

File: fml/raster_thread_merger_unittests.cc [hunk 5]
```diff
@@ -301,3 +301,143 @@ TEST(RasterThreadMerger, HandleTaskQueuesAreTheSame) {
...
+
+  latch1.Wait();
+  latch2.Wait();
+
+  fml::TaskQueueId qid1 = loop1->GetTaskRunner()->GetTaskQueueId();
+  fml::TaskQueueId qid2 = loop2->GetTaskRunner()->GetTaskQueueId();
+  const auto raster_thread_merger_ =
+      fml::MakeRefCounted<fml::RasterThreadMerger>(qid1, qid2);
+
+  raster_thread_merger_->Disable();
```
Comment (cyanglaz) [inline fml/raster_thread_merger_unittests.cc:334]: Can we add some assert after this line to ensure the raster_thread_merge is truely disabled? So we know `enable` worked.

File: shell/common/shell.cc [hunk 6]
```diff
@@ -708,6 +723,17 @@ void Shell::OnPlatformViewDestroyed() {
...
+  // Prevent any request to change the thread configuration for raster and
+  // platform queues while the platform view is being destroyed.
+  //
+  // This prevents false positives such as this method starts assuming that the
+  // raster and platform queues have a given thread configuration, but then the
+  // configuration is changed by a task, and the asumption is not longer true.
+  //
+  // This incorrect assumption can lead to dead lock.
+  // See `should_post_raster_task` for more.
+  rasterizer_->DisableThreadMergerIfNeeded();
```
Comment (cyanglaz) [inline shell/common/shell.cc:735]: Let's move this right before `should_post_raster_task`, so it's clear that `should_post_raster_task` is checked right after disabling thread merger.

File: shell/common/shell.cc [hunk 7]
```diff
@@ -619,6 +619,17 @@ void Shell::OnPlatformViewCreated(std::unique_ptr<Surface> surface) {
...
+  // Prevent any request to change the thread configuration for raster and
+  // platform queues while the platform view is being created.
+  //
+  // This prevents false positives such as this method starts assuming that the
+  // raster and platform queues have a given thread configuration, but then the
+  // configuration is changed by a task, and the asumption is not longer true.
+  //
+  // This incorrect assumption can lead to dead lock.
+  // See `should_post_raster_task` for more.
+  rasterizer_->DisableThreadMergerIfNeeded();
```
Comment (cyanglaz) [inline shell/common/shell.cc:631]: Let's move this right before `should_post_raster_task`, so it's clear that `should_post_raster_task` is checked right after disabling thread merger.

File: shell/common/shell.cc (hunk 6 above)
Comment (blasten) [reply shell/common/shell.cc:736]: Done.

File: shell/common/shell.cc (hunk 7 above)
Comment (blasten) [reply shell/common/shell.cc:631]: Done.

File: fml/raster_thread_merger_unittests.cc (hunk 5 above)
Comment (blasten) [reply fml/raster_thread_merger_unittests.cc:337]: Done.

**Pull request merged**